src/networkd.c

```diff
@@ -82,8 +82,18 @@ append_match_section(const NetplanNetDefinition* def, GString* s, gboolean match
         g_strfreev(split);
     } else if (def->match.driver)
         g_string_append_printf(s, "Driver=%s\n", def->match.driver);
-    if (def->match.mac)
-        g_string_append_printf(s, "MACAddress=%s\n", def->match.mac);
+    if (def->match.mac) {
+        /* LP: #1804861 and LP: #1888726:
+         * Using bond, bridge, and VLAN devices results in sharing MAC
+         * addresses across interfaces.  Match by PermanentMACAddress to match
+         * only the real phy interface and to continue to match it even after
+         * its MAC address has been changed.
+         */
+        if (def->bond || def->bridge || def->has_vlans)
```

Comment:
The approach looks reasonable. I wonder though if there is ever a use case where specifying `PermamentMacAddress` is incorrect. If there isn't then your proposal could be simplified significantly since it would simple replace `MacAddress` with `PermanentMacAddress`.

Reply:
> I wonder though if there is ever a use case where specifying `PermamentMacAddress` is incorrect.

Right, I wonder that too. I couldn't think of a use case where it would be incorrect.

But I wasn't sure how much risk the netplan project was willing to take. If we just replace `MACAddress` with `PermanentMACAddress` and there is some scenario where it matters, it will break. So I was trying to err on the conservative side.

Reply:
Yes, that makes sense. I wasn't able to find anything so far suggesting whether it would be safe to always use `PermanentMacAddress` or not. A safer path forward might be to add a `permanentmacaddress` option to netplan so that you could write:
```yaml
network:
    ethernets:
        ens3:
            dhcp4: true
            match:
                permanentmacaddress: 52:54:00:37:14:a4
            set-name: ens3
    version: 2
```
In other words, instead of adding some logic that would try to pick which MAC address to use, the above would leave that decision up to the user.

Reply:
That would be the safest, yes. That does impose some burden on the user to know which to use (and the documentation to try to explain the difference). In my case, I don't particularly care, as I've gone to all the work to figure this out. So I'd be happy to use either. For me, the main thing is that I need _some_ solution, as this is blocking my use of the offload options, which is forcing me to have a dispatcher hook script rather than everything being declarative syntax.